internal/uploader/events_uploader.go

```diff
@@ -0,0 +1,261 @@
+	if err := gz.Close(); err != nil {
+		return nil, errors.Wrapf(err, "failed closing gzip file")
+	}
+	return buffer, nil
+}
+
+func eventsFile(ctx context.Context, clusterID *strfmt.UUID, eventsHandler eventsapi.Handler, tw *tar.Writer) error {
+	if eventsHandler == nil {
+		return errors.Errorf("failed to get events for cluster %s, events handler is nil", clusterID)
+	}
+	dbEvents, err := eventsHandler.V2GetEvents(ctx, clusterID, nil, nil, models.EventCategoryMetrics, models.EventCategoryUser)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find events for cluster %s", *clusterID)
+	}
+
+	var events []*models.Event
+	for _, dbEvent := range dbEvents {
+		events = append(events, &dbEvent.Event)
+	}
+
+	contents, err := json.MarshalIndent(events, "", " ")
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal events")
+	}
+	return addFile(tw, contents, fmt.Sprintf("%s/events.json", clusterID))
+}
+
+func clusterFile(db *gorm.DB, tw *tar.Writer, clusterID *strfmt.UUID, pullSecret *serializedAuth) (*common.Cluster, error) {
+	if clusterID != nil {
+		cluster, err := common.GetClusterFromDBWithHosts(db, *clusterID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get cluster %s from db", *clusterID)
+		}
+		// To distinguish who is uploading the data
+		if cluster.EmailDomain == "" {
+			cluster.EmailDomain = getEmailDomain(pullSecret.Email)
+		}
+		if cluster.UserName == "" {
+			if token, tokenErr := ensureTokenFormat(pullSecret.Auth); tokenErr == nil {
+				cluster.UserName = getUsernameFromToken(token)
+			}
+		}
+
+		clusterJson, err := json.Marshal(cluster.Cluster)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to marshal cluster %s", *clusterID)
+		}
+		return cluster, addFile(tw, clusterJson, fmt.Sprintf("%s/events/cluster.json", *clusterID))
+	}
+	return nil, errors.New("no cluster id provided for cluster file")
+}
+
+func infraEnvFile(db *gorm.DB, tw *tar.Writer, infraEnvID *strfmt.UUID, clusterID *strfmt.UUID) error {
+	if infraEnvID != nil {
+		infraEnv, err := common.GetInfraEnvFromDB(db, *infraEnvID)
+		if err != nil {
+			return errors.Wrapf(err, "error getting infra-env %s from db", *infraEnvID)
+		}
+		iJson, err := json.Marshal(infraEnv.InfraEnv)
+		if err != nil {
+			return errors.Wrapf(err, "error marshalling infra-env")
+		}
+		// add the file content
+		return addFile(tw, iJson, fmt.Sprintf("%s/events/infraenv.json", *clusterID))
+	}
+	return errors.New("no infra-env id provided for infra-env file")
+}
+
+func hostsFile(db *gorm.DB, tw *tar.Writer, cluster *common.Cluster) (*strfmt.UUID, error) {
+	if cluster == nil {
+		return nil, errors.New("no cluster specified for hosts file")
+	}
+	var infraEnvID *strfmt.UUID
+	hosts := cluster.Cluster.Hosts
+	if hosts != nil {
+		hostJson, err := json.MarshalIndent(hosts, "", " ")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed marshalling hosts for cluster %s into file", *cluster.ID)
+		}
+		for _, host := range hosts {
```

Comment:
funk.ForEach could look a bit nicer for this

Reply:
Thanks for the suggestion! I'm not sure I fully understand, so how does this look?
```
		funk.ForEach(hosts, func(host *models.Host){
			if host.InfraEnvID.String() != "" {
				infraEnvID = &host.InfraEnvID
				return
			}
		})
```

Reply:
Wow that does look a lot cleaner, thank you! I'll modify this